packages/sol/src/availabilitySlots.ts

```diff
@@ -0,0 +1,148 @@
+import { IAvailabilitySlot } from "@litespace/types";
+import { dayjs } from "@/dayjs";
+import { flatten, orderBy } from "lodash";
+import { Schedule } from "./rule";
+import { start } from "repl";
+
+/**
+  * Divide list of slots into sub-slots. It is used whenever a user wants to book a lesson
+  * e.g., slot from 8am-10pm, split by 30min => 8-8:30, 8:30-9, 9-9:30, 9:30-10.
+  */
+export function getSubSlotsBatch(
+  /**
+   * list of the slots to be divided.
+   */
+  slots: IAvailabilitySlot.Slot[], 
+  /**
+   * the duration of each subslot in minutes.
+   */
+  duration: number,
+): IAvailabilitySlot.SubSlot[] {
+  return flatten(slots.map(slot => getSubSlots(slot, duration)));
+}
+
+export function getSubSlots(
+  slot: IAvailabilitySlot.Slot,
+  duration: number,
+): IAvailabilitySlot.SubSlot[] {
+  let start = dayjs.utc(slot.start);
+  let end = start.add(duration, "minutes");
+  const subslots: IAvailabilitySlot.SubSlot[] = [];
+  while (end.isBefore(dayjs.utc(slot.end)) || end.isSame(dayjs.utc(slot.end))) {
+    subslots.push({
+      parent: slot.id,
+      start: start.toISOString(),
+      end: end.toISOString(),
+    });
+    start = end;
+    end = start.add(duration, "minutes");
+  }
+  return subslots;
+}
+
+/** 
+  * checks if the passed book info can be booked in the given slot.
+  */
+export function canBook({ 
+  /**
+   * the slot to be partailly or fully booked.
+   */
+  slot,
+  /**
+   * subslots that are already booked (to be subtracted).
+   */
+  bookedSubslots,
+  /**
+   * the time that's needed to be booked.
+   */
+  bookInfo,
+}: {
+   slot: IAvailabilitySlot.Slot;
+   bookedSubslots: IAvailabilitySlot.SubSlot[];
+   bookInfo: {
+     start: string;
+     duration: number; // in minutes
+   }
+}): boolean {
+  const availableSubSlots = subtractSlots({
+    slots: [slot], 
+    subslots: bookedSubslots,
+  });
+  const found = availableSubSlots.find(subslot => isSuperSlot(subslot, {
+    parent: subslot.parent,
+    start: dayjs.utc(bookInfo.start).toISOString(),
+    end: dayjs.utc(bookInfo.start).add(bookInfo.duration, "minutes").toISOString(),
+  }))
+  return !!found;
+}
+
+/**
+ * removes the booked slots (represented by subslots) from the `slots` and 
+ * then returns the free remaining space/time represented by subslots. 
+ */
+export function subtractSlots({
+  slots,
+  subslots,
+}: {
+   slots: IAvailabilitySlot.Slot[];
+   subslots: IAvailabilitySlot.SubSlot[];
+}): IAvailabilitySlot.SubSlot[] {
+  const res: IAvailabilitySlot.SubSlot[] = [];
+  for (const slot of slots) {
+    const filtered = subslots.filter(subslot => isSuperSlot(slot, subslot));
+    res.push(
+      ...Schedule.splitBy(slot, filtered).map(event => ({
```

Comment:
`Schedule` belongs to the old logic (rules) and it will be removed once we move to slots. Please write a similar logic here using our new terminology.